### PR TITLE
idleTimeout in HTTP

### PR DESCRIPTION
**Pull request by ktoso**
I'm not sure it covers all the cases yet

Behaviour is:

Server:

```
[DEBUG] [10/23/2015 17:17:49.873] [ServerTest-akka.actor.default-dispatcher-3] [akka://ServerTest/system/IO-TCP/selectors/$a/0] New connection accepted
timeout = 5 seconds
[DEBUG] [10/23/2015 17:17:49.935] [ServerTest-akka.io.pinned-dispatcher-5] [akka.serialization.Serialization(akka://ServerTest)] Using serializer[akka.serialization.JavaSerializer] for message [akka.io.SelectionHandler$ChannelReadable$]
key = IdleTimeoutCheckTimer until 5 seconds // my debug println
... 
key = IdleTimeoutCheckTimer until 5 seconds
key = IdleTimeoutCheckTimer until 5 seconds
[DEBUG] [10/23/2015 17:17:55.208] [ServerTest-akka.actor.defa

ult-dispatcher-3] [akka.serialization.Serialization(akka://ServerTest)] Using serializer[akka.serialization.JavaSerializer] for message [akka.actor.Terminated]
[ERROR] [10/23/2015 17:17:55.209] [ServerTest-akka.actor.default-dispatcher-14] [ActorSystem(ServerTest)] Internal server error, sending 500 response
java.util.concurrent.TimeoutException: No elements passed in the last 5 seconds.
    at akka.stream.io.Timeouts$IdleTimeout$$anon$4.onTimer(Timeouts.scala:122)
    at akka.stream.stage.GraphStageLogic.akka$stream$stage$GraphStageLogic$$onInternalTimer(GraphStage.scala:323)
    at akka.stream.stage.GraphStageLogic$$anonfun$akka$stream$stage$GraphStageLogic$$getTimerAsyncCallback$1.apply(GraphStage.scala:101)
    at akka.stream.stage.GraphStageLogic$$anonfun$akka$stream$stage$GraphStageLogic$$getTimerAsyncCallback$1.apply(GraphStage.scala:101)
    at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:355)
    at akka.actor.Actor$class.aroundReceive(Actor.scala:467)
    at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:291)
    at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
    at akka.actor.ActorCell.invoke(ActorCell.scala:487)
    at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)
    at akka.dispatch.Mailbox.run(Mailbox.scala:220)
    at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

client:

```
$ REQ="POST / HTTP/1.1\r\nHost: 127.0.0.1:8080\r\nAccept: *\r\nContent-Length: 17\r\n"
ktoso @ 月~/code/akka [wip-idleTimeout-ktoso*]
$ (echo -ne $REQ; sleep 10;) | netcat localhost 8080
read(net): Connection reset by peer
```

capturing_from_loopback__lo0

It can/will be cleaned up a bit (want to add more tests), but general place where to put it seems OK.

### CODE REVIEW

**Comment by ktoso** [comment]
Resolves #18763 #16597 
// cc @jrudolph @sirthias

**Comment by ktoso** [comment]
Notes: 
- `Timeouts.asCompletion` is currently not used, I thought it should have been but then it's the wrong timeout I guess, the Timeout stages should throw a custom one. Happy to delete or adjust it.
- We could make the error nicer, but that needs a custom exception type in Timeouts anyway - then we can lift it into "HTTP IdleTimeout exception" which explains it more in the context of http.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Tcp.scala [hunk 1]
```diff
@@ -206,7 +203,10 @@ class Tcp(system: ExtendedActorSystem) extends akka.actor.Extension {
                          connectTimeout: Duration = Duration.Inf,
                          idleTimeout: Duration = Duration.Inf): Flow[ByteString, ByteString, Future[OutgoingConnection]] = {
 
-    val remoteAddr = remoteAddress
+    val timeoutHandling = idleTimeout match {
+      case d: FiniteDuration ⇒ Flow[ByteString].via(Timeouts.idleTimeout(d))
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Tcp.scala]: this should be `join(Timeouts.idleTimeoutBidi)`. Otherwise this will timeout on a long response, which is not idle at all.

File: akka-stream/src/main/scala/akka/stream/io/Timeouts.scala [hunk 2]
```diff
@@ -168,4 +173,13 @@ object Timeouts {
     }
   }
 
+  private final val failureAsCompletionFlow = Flow[Any].transform(() ⇒ new PushStage[Any, Any] {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/io/Timeouts.scala]: why is this needed? This feels wrong.

File: akka-stream/src/main/scala/akka/stream/impl/io/TcpListenStreamActor.scala [hunk 3]
```diff
@@ -151,10 +155,17 @@ private[akka] class TcpListenStreamActor(localAddressPromise: Promise[InetSocket
     val (connected: Connected, connection: ActorRef) = incomingConnections.dequeueInputElement()
     val tcpStreamActor = context.watch(context.actorOf(TcpStreamActor.inboundProps(connection, halfClose, settings)))
     val processor = ActorProcessor[ByteString, ByteString](tcpStreamActor)
+
+    import scala.concurrent.duration._
+    val handler = (idleTimeout match {
+      case d: FiniteDuration ⇒ Flow[ByteString].via(Timeouts.idleTimeout(d))
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpListenStreamActor.scala]: this should use the BidiStage version, this only checks for the input side, and does not consider long responses.

File: akka-stream/src/main/scala/akka/stream/io/Timeouts.scala (hunk 2 above)
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/io/Timeouts.scala]: > Timeouts.asCompletion is currently not used, I thought it should have been but then it's the wrong timeout I guess, the Timeout stages should throw a custom one. Happy to delete or adjust it.

https://github.com/akka/akka/pull/18770#issuecomment-150613936

Yeah, I'll remove it. I don't think it's wrong per-se, but not something we need to provide (nor is used).

File: akka-stream/src/main/scala/akka/stream/impl/io/TcpListenStreamActor.scala (hunk 3 above)
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpListenStreamActor.scala]: Fixed, thanks for the reminder!

File: akka-stream/src/main/scala/akka/stream/scaladsl/Tcp.scala (hunk 1 above)
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Tcp.scala]: Thanks, fixed now

**Comment by ktoso** [comment]
Amended and pushed.

**Comment by drewhk** [comment]
LGTM

File: akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala [hunk 4]
```diff
@@ -138,6 +139,42 @@ class ClientServerSpec extends WordSpec with Matchers with BeforeAndAfterAll {
...
+      // await for the server to get the request
+      val serverReceivedRequestAtNanos = Await.result(receivedRequest.future, 2.seconds)
+
+      // waiting for the timeout to happen on the client
+      Try(Await.result(clientsResponseFuture, 2.second)).recoverWith {
+        case _: StreamTcpException ⇒ Success(System.nanoTime())
+        case other: Throwable      ⇒ Failure(other)
+      }.get
+      val diff = System.nanoTime() - serverReceivedRequestAtNanos
+      diff should be > 300.nanoseconds.toNanos
```
Comment (viktorklang) [inline akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala]: 300ns is a veeeery short time

File: akka-stream/src/main/scala/akka/stream/impl/io/TcpConnectionStream.scala [hunk 5]
```diff
@@ -302,6 +304,7 @@ private[akka] class InboundTcpStreamActor(
 private[akka] class OutboundTcpStreamActor(processorPromise: Promise[Processor[ByteString, ByteString]],
                                            localAddressPromise: Promise[InetSocketAddress],
                                            _halfClose: Boolean,
+                                           idleTimeout: Duration,
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpConnectionStream.scala]: Reason why it is not a FiniteDuration?

File: akka-stream/src/main/scala/akka/stream/impl/io/TcpListenStreamActor.scala [hunk 6]
```diff
@@ -151,10 +155,17 @@ private[akka] class TcpListenStreamActor(localAddressPromise: Promise[InetSocket
     val (connected: Connected, connection: ActorRef) = incomingConnections.dequeueInputElement()
     val tcpStreamActor = context.watch(context.actorOf(TcpStreamActor.inboundProps(connection, halfClose, settings)))
     val processor = ActorProcessor[ByteString, ByteString](tcpStreamActor)
+
+    import scala.concurrent.duration._
+    val handler = (idleTimeout match {
+      case d: FiniteDuration ⇒ Flow[ByteString].join(Timeouts.idleTimeoutBidi[ByteString, ByteString](d))
+      case _                 ⇒ Flow[ByteString]
+    }).andThenMat(() ⇒ (processor, ()))
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpListenStreamActor.scala]: Doesn't this mean that it can only be materialized once?

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpListenStreamActor.scala]: correct, but it's a connection - doesn't make much sense to materialize it twice.

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpListenStreamActor.scala]: I'm just worried that "hacks" such as this will become "common practice" amongst beginners.

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpListenStreamActor.scala]: Hm... this *is* tcp handling internals though, there is stuff like that on these levels - i.e. materializing the connection stream twice will not work, and we tell people about this.

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpListenStreamActor.scala]: Perhaps I was unclear: Don't you agree that having to do this *could* be a sign of lack in the API?

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpListenStreamActor.scala]: Oh, this way around... let's ticketify maybe? Can you open one?

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpListenStreamActor.scala]: My brain is fried right now :(

File: akka-stream/src/main/scala/akka/stream/impl/io/TcpConnectionStream.scala (hunk 5 above)
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpConnectionStream.scala]: We allow disabling this feature - the setting can be "off" in the config

File: akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala (hunk 4 above)
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala]: ouch, typo - millis

**Comment by ktoso** [comment]
Updated minor fixups, let's discus what @viktorklang mentioned in a separate ticket

File: akka-stream/src/main/scala/akka/stream/impl/io/TcpConnectionStream.scala (hunk 5 above)
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpConnectionStream.scala]: Ok, so anything non-finite == off then I presume

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpConnectionStream.scala]: correct

File: akka-stream/src/main/scala/akka/stream/impl/io/TcpListenStreamActor.scala (hunk 6 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpListenStreamActor.scala]: I am not sure what is expected here. This is an *incoming* TCP connection, so there is no way you can materialize it twice, since the connection is created by the client. We discussed with @rkuhn to make it reusable by implicitly adding merge and broadcast logics to the input/output sides of the connection, but that in my opinion is much worse as it can lead to silent data corruption on the wire (if someone is not aware of the merging) while the current approach throws an exception instead.

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpListenStreamActor.scala]: btw, you can use `.via(Flow.fromProcessor(processor))`, because `andThenMat` is going away (it is an internal API).

**Comment by ktoso** [comment]
Failure was unrelated, it was: https://github.com/akka/akka/issues/18756
